### ISSUE TITLE
Broken master-master on  box.cfg recall

### ISSUE THREAD

**Issue by Gerold103** [tarantool/tarantool#3636]
A description in broad lines: I call `box.cfg{replication = copy of box.cfg.replication}` on a cluster with slow disk, and get the "ER_CFG: Incorrect value for option 'replication': duplicate connection with the same replica UUID" error.
Test:
```Lua
-- Instance 1 --
-- Step 1 --
fiber = require('fiber')
box.cfg{
	listen = 3313,
	replication = {'localhost:3313', 'localhost:3314'}
}
box.schema.user.grant('guest', 'read,write,execute', 'universe')
-- Step 3 --
s = box.schema.create_space('test')
pk = s:create_index('pk')
box.error.injection.set("ERRINJ_WAL_DELAY", true)
f = fiber.create(function() box.space.test:replace{1, 1} end)
box.space.test:select{} -- The tuple is in memory, but not on disk.
-- Step 5 --
box.error.injection.set("ERRINJ_WAL_DELAY", false)


-- Instance 2 --
-- Step 2 --
fiber = require('fiber')
box.cfg{
	listen = 3314,
	replication = {'localhost:3313', 'localhost:3314'},
	read_only = true
}
while not box.space.test do fiber.sleep(0.01) end
while not box.space.test.index.pk do fiber.sleep(0.01) end
-- Step 4 --
-- Reconfigure to exactly the same replication.
box.cfg({replication = {'localhost:3313', 'localhost:3314'}})
```

After step 5 I see ER_CFG error. What is more, I see it **after** box.cfg is returned on instance 2.
Without writing a tuple the bug is not reproducing.

**Comment by Gerold103** [comment]
As I remember, the problem emerged after this commit: https://github.com/tarantool/tarantool/commit/c1a16b26062522f1c92f6af602e82edea7dd1785

**Comment by GeorgyKirichenko** [comment]
This is by design and couldn't be fixed without behaviour changes

Replica closes existing applier and then creates a new one to the same master.
Master receives a socket close event in a relay thread and a subscribe request in tx thread that might lead to a race (networking and threading)

Possible solutions:
1. introduce disconnect and disconnect request into a replication proto
2. move relay to the tx thread (but this doesn't solve a networking race)
3. reuse applier if it has the same uri and OK status

**Comment by Gerold103** [comment]
Why do you need my feedback and how does the comment above fix the bug? Regardless of reasons, the bug exists so the issue as well should not be closed.

**Comment by locker** [comment]
Fixed by 4baa71bcb82e690776b5dd7f273d685c40260712

**Comment by locker** [comment]
The issue was fixed in 1.10. Here's the log after the second call to box.cfg:
```
tarantool> box.cfg{
         > listen = 3313,
         > replication = {'localhost:3313', 'localhost:3314'},
         > read_only = true,
         > }
2018-10-03 15:29:54.408 [15377] main/101/interactive I> set 'read_only' configuration option to true
2018-10-03 15:29:54.408 [15377] main/101/interactive I> connecting to 2 replicas
2018-10-03 15:29:54.410 [15377] main/117/applier/localhost:3313 I> remote master is 1.10.1 at 127.0.0.1:3313
2018-10-03 15:29:54.410 [15377] main/118/applier/localhost:3314 I> remote master is 1.10.1 at 127.0.0.1:3314
2018-10-03 15:29:54.410 [15377] main/101/interactive I> connected to 2 replicas
2018-10-03 15:29:54.411 [15377] main/101/interactive I> synchronizing with 2 replicas
2018-10-03 15:29:54.411 [15377] main/118/applier/localhost:3314 I> can't join/subscribe
2018-10-03 15:29:54.411 [15377] main/118/applier/localhost:3314 xrow.c:892 E> ER_CFG: Incorrect value for option 'replication': duplicate connection with the same replica UUID
2018-10-03 15:29:54.411 [15377] main/118/applier/localhost:3314 I> will retry every 1.00 second
2018-10-03 15:29:55.416 [15377] main/101/interactive I> replica set sync complete
2018-10-03 15:29:55.416 [15377] main/101/interactive I> set 'replication' configuration option to ["localhost:3313","localhost:3314"]
---
...

tarantool> 2018-10-03 15:29:55.520 [15377] relay/127.0.0.1:49778/101/main C> exiting the relay loop
2018-10-03 15:29:56.524 [15377] relay_0x55ab2f7941b0/101/main I> recover from `./00000000000000000000.xlog'
2018-10-03 15:29:56.524 [15377] relay_0x55ab2f7941b0/101/main I> done `./00000000000000000000.xlog'

---
...

tarantool>
```

Note, an error message is still printed to the log, because the old relay is running on the master when we call box.cfg and we can't do anything about it, as Georgy pointed out. However, the error is transient: the applier says that it will keep trying to connect and in a second it does connect and successfully completes synchronization so that the instance doesn't enter the oprhan mode ("failed to synchronize with 1 out of 2 replicas" on 1.9 vs "replica set sync complete" on 1.10).
